File: 2022/task2/task2.js
const readline = require('readline').createInterface(process.stdin, process.stdout);

let width = null;
let height = null;
const map = [];
const queue = [];
let linesCount = 0;
let nextRegionIndex = 0;
const regions = [];

const getRegionArea = (region) => (region.maxX - region.minX + 1) * (region.maxY - region.minY + 1);
const getRegionValue = (region) => {
    let goodSpots = 0;

    for (let x = region.minX; x <= region.maxX; x++) {
        for (let y = region.minY; y <= region.maxY; y++) {
            goodSpots += map[y][x].value;
        }
    }

    return goodSpots;
}

const checkNeighbours = (centerX, centerY, point) => {
    for (let x = centerX - 1; x <= centerX + 1; x++) {
        if (x < 0 || x >= width) {
            continue;
        }
        for (let y = centerY - 1; y <= centerY + 1; y++) {
            if (y < 0 || y >= height) {
                continue;
            }

            if (map[y][x].value) {
                map[y][x].region = point.region;
                queue.unshift({ x, y })
            }
        }
    }
}

const getBestRegionSize = () => {
    while (queue.length > 0) {
        const candidate = queue[0];
        let currentX = candidate.x;
        let currentY = candidate.y;
        const point = map[currentY][currentX];

        if (point.region === null) {
            point.region = nextRegionIndex;
            regions.push({ minX: currentX, maxX: currentX, minY: currentY, maxY: currentY });
            nextRegionIndex += 1;
        }
        let region = regions[point.region];

        if (point.visited) {
            queue.shift();
            continue;
        }

        point.visited = true;

        if (point.value) {
            region.minX = Math.min(currentX, region.minX);
            region.maxX = Math.max(currentX, region.maxX);
            region.minY = Math.min(currentY, region.minY);
            region.maxY = Math.max(currentY, region.maxY);
            checkNeighbours(currentX, currentY, point);
        }
    }

    const resultRegions = regions
        .filter(item => getRegionArea(item) > 1)
        .map(region => ({
            value: getRegionValue(region),
            area: getRegionArea(region),
        }))
        .sort((a, b) => {
            const valueDiff = (b.value / b.area) - (a.value / a.area);
            return valueDiff === 0 ? b.area - a.area : valueDiff;
        });

    return resultRegions.length > 0 ? resultRegions[0].area : 0;
};

readline.on('line', (line) => {
    if (linesCount === 0) {
        [width, height] = line.split(' ').map(item => Number(item));
    } else {

        map.push(line.split(' ').map((item, index) => {
            if (item === '1') {
                queue.push({ x: index, y: linesCount - 1 })
            }
            return {
                value: Number(item),
                visited: false,
                region: null,
            }
        }));

    }
    if (linesCount++ === height) {
        console.log(getBestRegionSize());
        readline.close();
    }
}).on('close', () => process.exit(0));

File: 2022/task2/task2.py
n, m = map(int, input().split())
data_map = [[0] * (n + 2)]
for _ in range(m):
    data_map.append([0] + list(map(int, input().split())) + [0])
data_map.append([0] * (n + 2))

good_spots = set()
for i in range(m + 2):
    for j in range(n + 2):
        if data_map[i][j] == 1:
            good_spots.add((i, j))

regions = []
while good_spots:
    region_spots = [good_spots.pop()]
    max_x = 1
    max_y = 1
    min_x = m + 2
    min_y = m + 2
    while region_spots:
        x, y = region_spots.pop(-1)
        max_x = max(max_x, x)
        min_x = min(min_x, x)
        max_y = max(max_y, y)
        min_y = min(min_y, y)

        for step_x, step_y in ((-1, -1), (-1, 0), (-1, +1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1)):
            new_x = x + step_x
            new_y = y + step_y
            if (new_x, new_y) in good_spots:
                region_spots.append((new_x, new_y))
                good_spots.remove((new_x, new_y))

    regions.append((min_x, min_y, max_x, max_y))

efficiency = 0
best_region_area = 0

for x1, y1, x2, y2 in regions:
    region_area = (x2 - x1 + 1) * (y2 - y1 + 1)
    region_good_spots_count = 0
    for row in range(x1, x2 + 1):
        region_good_spots_count += sum(data_map[row][y1:y2 + 1])

    if region_good_spots_count > 1:
        efficiency_temp = region_good_spots_count / region_area
        if efficiency_temp > efficiency:
            efficiency = efficiency_temp
            best_region_area = region_area
        elif efficiency_temp == efficiency:
            if region_area > best_region_area:
                efficiency = efficiency_temp
                best_region_area = region_area

print(best_region_area)
